Update to Jose 6

mbeidler opened the pull request:
This resolves #59. It also updates the LTS resolver and servant dependencies (to the latest).  It is necessary to change them to a later version to get it to compile with Jose 6, but let me know if I haven't selected the optimal versions. I ran the tests and they passed.

mbeidler commented:
Sorry for all the commits, either I can clean them up or squash on merge.

mbeidler commented:
Sorry for all the commits, either I can clean them up or squash on merge.

mchaver commented:
@mbeidler I will squash and merge it is ready, can you fix the tests?

mbeidler commented:
@mchaver I did a little more digging and the build is failing due to a bug in the haddocks version in later LTS snapshots.

Previous to this PR, `servant-auth` was using 6.7
```bash
stack exec --resolver=lts-6.7 -- haddock --version
Haddock version 2.16.1, (c) Simon Marlow 2006
```

This PR modifies it to use 9.6:
```bash
stack exec --resolver=lts-9.6 -- haddock --version
Haddock version 2.17.3, (c) Simon Marlow 2006
```

And **2.17.3** is the version that has the bug.

A temporary fix until the updated haddocks is available in an LTS snapshot is to run stack haddock with a nightly. I tested the following and it completed successfully.

```bash
stack haddock --resolver nightly-2017-10-13
```

I'll add a commit to **.travis.yml** to use that.

Also, to see the working version the nightly uses:
```bash
stack exec --resolver nightly-2017-10-13 -- haddock --version
Haddock version 2.18.1, (c) Simon Marlow 2006
```

mchaver commented:
@mbeidler Thank you for investigating that issue. I took a quick look at the changes. I will merge this on Monday.

phadej commented on .travis.yml:
```diff
@@ -2,7 +2,6 @@ sudo: false
 
 env:
   - STACK_YAML=stack.yaml
-  - STACK_YAML=stack-ghc-8.0.2.yaml
```
We really want to test on 8.0.2. Please revert.

mbeidler replied:
The reason I removed that is because this PR updates servant-auth to use the 9.6 LTS snapshot which uses 8.0.2 (`stack.yaml`). Since that is the case, do we still need this?

mbeidler replied:
Ah, I think @jkarni answered my question below. I'll create a `stack-ghc-7.10.3.yaml`.

phadej commented on .travis.yml:
```diff
@@ -20,7 +19,9 @@ script:
   - stack setup --no-terminal
   - stack build --ghc-options=-Werror --no-terminal
   - stack test --ghc-options=-Werror --no-terminal
-  - stack haddock --no-terminal
```
rather add a `stack.yaml` for nightly and `if [ $STACK_YAML = "stack-nightly.yaml" ]; ...`

jkarni replied:
Agreed.

mbeidler replied:
@phadej @jkarni Currently `stack haddock` does not work unless it's using the nightly. Do you mean to only run `stack haddock` for `stack-nightly.yaml`?

phadej replied:
yes

mbeidler replied:
I noticed that the `stack-ghc-8.2.1.yaml` was already using a nightly so my latest commit just runs `stack haddock` for that.

jkarni commented on stack.yaml:
```diff
@@ -6,22 +6,22 @@ packages:
 - unix-time-0.3.7
 - wai-extra-3.0.19
 - warp-3.2.9
-resolver: lts-6.7
```
We do need CI (and stack files) for all GHCs from 7.10 till 8.2. I think it makes sense for the default `stack.yaml` to be on the latest released GHC (8.0), but we then need `stack-ghc-7.10.3.yaml` (which would essentially be this file before the changes.

jkarni commented on .travis.yml:
```diff
@@ -20,7 +19,9 @@ script:
   - stack build --ghc-options=-Werror --no-terminal
   - stack test --ghc-options=-Werror --no-terminal
-  - stack haddock --no-terminal
+  # temporarily use a nightly for haddocks until fix is available in LTS
```
Can you link to the haddock issue here so future us-es can keep tabs on when it's resolved (and what the issue is)?

mbeidler replied:
Sure. Will do.

mbeidler commented:
@jkarni @phadej I added the stack-ghc-7.10.3.yaml as suggested, however, I still need clarification on the the suggestion in the PR review. Is the suggestion to add a `stack-nightly.yaml` and only run `stack haddock` for that env?

mbeidler commented:
@mchaver @jkarni @phadej The Travis build is failing with "We couldn't find the repository 
plow-technologies/servant-auth".

mbeidler commented:
@phadej @jkarni I cleaned up the commit history with a rebase and I think I've addressed your feedback. Note that I noticed that `stack-ghc-8.2.1.yaml` is using a nightly build, so I didn't add an additional nightly file, but modified `.travis.yml` as suggested to only run `stack haddock` for that build. Let me know if there's anything else you want me to change.

tommyengstrom commented:
Hey guys. It would be awesome if this would get merged. Is there anything I can do to help?

mchaver commented:
@phadej @jkarni Anything else you guys want to see?

bergey commented:
FWIW, I've been using this at work for a few weeks, and AFAICT it works well.  It would be great to have a release number instead of a git commit hash. :)

phadej commented:
merged as #66
